packages/amplify-category-function/src/provider-utils/awscloudformation/utils/addLayerToFunctionUtils.ts

```diff
@@ -58,33 +60,55 @@ export const askLayerSelection = async (
+  let layerSelections: string[] = (await inquirer.prompt(layerSelectionQuestion)).layerSelections;
   const askArnQuestion = layerSelections.includes(provideExistingARNsPrompt);
   layerSelections = layerSelections.filter(selection => selection !== provideExistingARNsPrompt);
 
-  for (let selection of layerSelections) {
-    const currentSelectionDefaults = filterProjectLayers(previousSelections).find(sel => sel.resourceName === selection);
-    const currentVersion = currentSelectionDefaults ? currentSelectionDefaults.version.toString() : undefined;
-    const layerState = layerMetadataFactory(selection);
-    await layerState.syncVersions(); // make sure we are reflecting the latest changes;
-    const layerVersionPrompt: ListQuestion = {
-      type: 'list',
-      name: 'versionSelection',
-      message: versionSelectionPrompt(selection),
-      choices: layerState.listVersions().map(num => num.toString()),
-      default: currentVersion,
-      filter: numStr => parseInt(numStr, 10),
-    };
+  for await (const layerName of layerSelections) {
+    const layerCloudState = LayerCloudState.getInstance();
+    const layerVersions = await layerCloudState.getLayerVersionsFromCloud(context, layerName);
+    const layerVersionChoices = layerVersions.map(mapVersionNumberToChoice);
+
+    // skip asking version for a new layer
+    if (layerVersionChoices.length > 0) {
+      layerVersionChoices.unshift(defaultLayerVersionPrompt);
+      const previousLayerSelection = _.first(filterProjectLayers(previousSelections).filter(prev => prev.resourceName === layerName));
+
+      let defaultLayerSelection: string;
+      if (previousLayerSelection === undefined || previousLayerSelection.isLatestVersionSelected) {
+        defaultLayerSelection = defaultLayerVersionPrompt;
+      } else {
+        defaultLayerSelection = mapVersionNumberToChoice(_.first(layerVersions.filter(v => v.Version === previousLayerSelection.version)));
+      }
+
+      const versionSelection = (
+        await inquirer.prompt(layerVersionQuestion(layerVersionChoices, versionSelectionPrompt(layerName), defaultLayerSelection))
+      ).versionSelection;
+
+      const isLatestVersionSelected = versionSelection === defaultLayerVersionPrompt;
+      const selectedVersion =
+        versionSelection === defaultLayerVersionPrompt ? defaultLayerVersionPrompt : Number(_.first(versionSelection.split(':')));
+
+      lambdaLayers.push({
+        type: 'ProjectLayer',
+        resourceName: layerName,
+        version: selectedVersion,
+        isLatestVersionSelected: isLatestVersionSelected,
+        env: context.amplify.getEnvInfo().envName,
+      });
+    } else {
+      lambdaLayers.push({
+        type: 'ProjectLayer',
+        resourceName: layerName,
+        version: defaultLayerVersionPrompt,
+        isLatestVersionSelected: true,
+        env: context.amplify.getEnvInfo().envName,
+      });
```

Comment:
There is a decent amount of duplication here. Can we call `push()` in one place and use the branching logic to only compute the fields that are different.